amaranth/hdl/mem.py

```diff
@@ -254,6 +289,8 @@ class WritePort(Elaboratable):
     divide memory width evenly.
     """
     def __init__(self, memory, *, domain="sync", granularity=None, src_loc_at=0):
+        if granularity is not None and isinstance(memory.shape, ShapeCastable) or memory.shape.signed:
```

Comment:
I think a check for `not (isinstance(memory.shape, Shape) and memory.shape.unsigned)` is clearer.